Storage/SQL/SQLiteLogins.swift

```diff
@@ -0,0 +1,181 @@
+
+        if let timeCreated = row.getTimestamp("timeCreated"),
+            let timeLastUsed = row.getTimestamp("timeLastUsed"),
+            let timePasswordChanged = row.getTimestamp("timePasswordChanged") {
+                login.timeCreated = timeCreated
+                login.timeLastUsed = timeLastUsed
+                login.timePasswordChanged = timePasswordChanged
+        }
+
+        return login
+    }
+
+    private class func LoginDataFactory(row: SDRow) -> LoginData {
+        return LoginFactory(row) as LoginData
+    }
+
+    private class func LoginUsageDataFactory(row: SDRow) -> LoginUsageData {
+        return LoginFactory(row) as LoginUsageData
+    }
+
+    public func getLoginsForProtectionSpace(protectionSpace: NSURLProtectionSpace) -> Deferred<Result<Cursor<LoginData>>> {
+        let sql = "SELECT username, password, hostname, httpRealm, formSubmitUrl, usernameField, passwordField FROM \(table.name) WHERE hostname = ? ORDER BY timeLastUsed DESC"
+        let args: [AnyObject?] = [protectionSpace.host]
+        return db.runQuery(sql, args: args, factory: SQLiteLogins.LoginDataFactory)
+    }
+
+    public func getUsageDataForLogin(login: LoginData) -> Deferred<Result<LoginUsageData>> {
+        let sql = "SELECT * FROM \(table.name) WHERE hostname = ? AND username IS ? LIMIT 1"
+        let args: [AnyObject?] = [login.hostname, login.username]
+        return db.runQuery(sql, args: args, factory: SQLiteLogins.LoginUsageDataFactory) >>== { value in
+            return deferResult(value[0]!)
+        }
+    }
+
+    public func addLogin(login: LoginData) -> Success {
+        var args = [AnyObject?]()
+        args.append(login.hostname)
+        args.append(login.httpRealm)
+        args.append(login.formSubmitUrl)
+        args.append(login.usernameField)
+        args.append(login.passwordField)
+
+        if var login = login as? SyncableLoginData {
+            if login.guid == nil {
+                login.guid = NSUUID().UUIDString
+            }
+            args.append(login.guid)
+        } else {
+            args.append(NSUUID().UUIDString)
+        }
+
+        let date = NSNumber(unsignedLongLong: NSDate.nowMicroseconds())
+        args.append(date) // timeCreated
+        args.append(date) // timeLastUsed
+        args.append(date) // timePasswordChanged
+        args.append(login.username)
+        args.append(login.password)
+
+        return db.run("INSERT INTO \(table.name) (hostname, httpRealm, formSubmitUrl, usernameField, passwordField, guid, timeCreated, timeLastUsed, timePasswordChanged, username, password) VALUES (?,?,?,?,?,?,?,?,?,?,?)", withArgs: args)
+    }
+
+    public func addVisitFor(login: LoginData) -> Success {
+        let date = NSNumber(unsignedLongLong: NSDate.nowMicroseconds())
+        return db.run("UPDATE \(table.name) SET timeLastUsed = ? WHERE hostname = ? AND username IS ?", withArgs: [date, login.hostname, login.username])
```

Comment:
This would be nice if the `LoginData` could preserve its ID privately and use that in the `WHERE` clause. Worth a follow-up.